rfd/0002-custom-terraform-registry.md

```diff
@@ -0,0 +1,185 @@
+hosted externally.
+
+#### Package Hosting
+
+The distribution files (i.e. the `zip`, checksum and signature files) indexed
+by the registry will be stored in an AWS S3 bucket, and exposed to the Internet
+via a CloudFront CDN service.
+
+The storage bucket will be configured as per the Gold-class storage requirements
+described in the [Cloud Artifact Storage Standards](https://github.com/gravitational/cloud/blob/master/rfd/0017-artifact-storage-standards.md)
+RFD.
+
+#### Why can't we just use Houston?
+
+The existing Houston distribution tool at [get.gravitational.com](https://get.gravitational.com/)
+is not suitable for distributing the files indexed by the registry. Houston has
+_very_ particular requirements about the naming of the files it serves (e.g.
+all files are expected to be `*.tar.gz` compressed tarballs - or at least
+_named_ as such) and, as it stands, will not serve the files we need it to.
+
+We _could_ modify Houston to serve the required files, but as the service is
+already slated for retirement, that does not seem to be a good use of time
+& money.
+
+### Building provider packages
+
+As far as the registry protocol is concerned, the entire package for a given
+provider consists of:
+
+* The actual binary distribution zipfile
+* A checksum file containing the SHA256 of the zipfile (formatted as per
+ `sha256sum` tool)
+* A signature file containing a binary, detached GPG signature for
+  checksum file.
+
+The Teleport provider release process currently produces a compressed tarball
+by default, and this should still be considered the primary artefact produced
+by the build, at least as long as Houston is expected to serve downloads of
+the Terraform provider.
+
+By analogy with the main Teleport release system, the registry-compatible
+provider package will be created during release promotion, with a build task
+that will:
+
+1. For each release tarball (one each for multiple release platforms):
+    1. repack the tarball into an appropriately formatted & named zipfile
+    2. generate the checksum file from the zipfile
+    3. generate the signature file by signing the checksum file
+2. Upload all generated files to the AWS bucket that backs the registry
+3. Update the registry index to include the newly released providers (see
+   _Populating the registry_, below)
+
+I expect this to be implemented as a small Go program, rather than a shell
+script. Using small Go programs for build tasks has shown itself to be a
+useful, flexible and above all _legible_ tool for scripting build tasks in
+the mainline Teleport repository, and we should continue that pattern in
+`teleport-plugins`.
+
+#### Build Security issues
+
+This process will require that the `teleport-plugins` promotion process will
+need new secrets:
+
+1. Read access to a signing key, to sign the package (see note below)
+2. Write access to the AWS bucket storing the signed packages for distribution
```

Comment:
Talking with @russjones, one of invariants we want to maintain is no secrets in CI. Does our Drone instance run in AWS? Can we use IAM roles instead of secrets?

Reply:
> invariants we want to maintain (or maybe rather, introduce and maintain) is no secrets in CI

We don't use Drone for teleport-plugins Continuous Integration anymore. We use GCB for CI and Drone for Release/CD. We do use Drone (with AWS secrets) for release processes, most of which this RFD is concerned with.

This invariant will hold -- no secrets in CI, because the secrets needed will be for release/CD.

> Does our Drone instance run in AWS?

It runs in GCP GKE -- in the `drone-gravitational` project.

> Can we use IAM roles instead of secrets?

We're still a way away from this.  We've got the right infra in AWS -- with well defined principle-of-least-privilege policies that can bet attached to roles/service users.  However the bootstrap to get a short lived secret/token ahead of any user controllable processes isn't available yet.

I'd be cautious about coupling the reslease secrets work with this terraform RFD. There is at least a RFD and weeks to months of work (e.g. machine ID for CD) to improve the state of the art for release secrets. IMO we don't want to slip on Q1 time to value work (this) for that security project that may make more sense to look at after machine ID is available and we look at using one of our dogfooding teleport clusters as an aws intermediary for CD.